### PR TITLE
Added dict support for optimizing

### PR DESCRIPTION
**Pull request by wsuchy**
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and open-source license: see below

#### Summary
I've just realized I forgot to add support dict for `data` and `inits` parameters which is already supported by sampling method.
#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

- Salesforce

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: cmdstanpy/cmdstan_args.py [hunk 1]
```diff
@@ -381,9 +381,12 @@ def validate(self) -> None:
                             ' found {}'.format(self.seed[i])
                         )
 
-        if self.data is not None:
+        if isinstance(self.data, str):
             if not os.path.exists(self.data):
                 raise ValueError('no such file {}'.format(self.data))
+        elif not isinstance(self.data, dict):
```
Comment (mitzimorris) [inline cmdstanpy/cmdstan_args.py]: don't you need to update line 293  - type hint on data - else how will self.data be a dict?

File: cmdstanpy/cmdstan_args.py [hunk 2]
```diff
@@ -381,9 +381,12 @@ def validate(self) -> None:
...
+            if isinstance(self.method_args, OptimizeArgs):
```
Comment (mitzimorris) [inline cmdstanpy/cmdstan_args.py]: include SamplerArgs as well here please

Comment (wsuchy) [reply cmdstanpy/cmdstan_args.py]: @mitzimorris I see in many `SamplerArgsTest` tests that data is null - is that expected to run sampling without any data file?

Comment (mitzimorris) [reply cmdstanpy/cmdstan_args.py]: yes, models without a data block are valid.

one use case is modelswhere the transformed data block is used to generate data for each chain/run via calls to the RNG functions.

Comment (wsuchy) [reply cmdstanpy/cmdstan_args.py]: I've fixed the logic, now we throw exception when data is None *and* we are optimizing, or when data is different than str or dict.

**Pull request merged**